reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,291 @@
+```
+
+The above command starts a Ray head node, it also starts a Ray autoscaler with
+above YAML configuration, it configures autoscaler with "ray_on_spark" `NodeProvider`
+backend, when Ray jobs are submitted Ray cluster is requested to scale up,
+the configured `NodeProvider` backend is called to trigger Ray worker node setup.
+
+The `NodeProvider` [interfaces](https://github.com/ray-project/ray/blob/e1c48ab8b1225daa44ea63ed1e7dc956b0e7f411/python/ray/autoscaler/node_provider.py#L13) are defined like following (only some important interfaces are listed here):
+
+```python
+
+class NodeProvider:
+
+    def create_node_with_resources_and_labels(
+        self, node_config, tags, count, resources, labels
+    ):
+        """
+        Create `count` number of nodes, resources are node CPU / GPU / memory
+        requirements that are defined in autoscaler config YAML file.
+        After node creation, generate a unique node_id and link the node id with
+        several node metadata like tags / labels.
+        """
+        ...
+    
+    def terminate_node(self, node_id):
+        """
+        Terminate node by provided node id.
+        """
+  
+    def external_ip(self, node_id):
+        """
+        Get node external ip. In Ray on spark NodeProvider implementation,
+        we need to use fake IP that equals to node id, see Key questions
+        "How to make `NodeProvider` backend support multiple Ray worker nodes
+        running on the same virtual machine"
+        """
+        ...
+
+    def internal_ip(self, node_id):
+        """
+        Get node internal ip. In Ray on spark NodeProvider implementation,
+        we need to use fake IP that equals to node id, see Key questions
+        "How to make `NodeProvider` backend support multiple Ray worker nodes
+        running on the same virtual machine"
+        """
+        ...
+
+
+    def is_running(self, node_id):
+        """
+        query node running status.
+        """
+        ...
+
+    def is_terminated(self, node_id):
+        """
+        query whether node is terminated
+        """
+        ...
+
+    def node_tags(self, node_id):
+        """
+        query node tags
+        """
+        ...
+
+    def set_node_tags(self, node_id, tags):
+        """
+        set node tags
+        """
+        ...
+
+    @staticmethod
+    def bootstrap_config(cluster_config):
+        """
+        Override original autoscaler configs here.
+        """
+        ...
+```
+
+### How to implement Ray on spark node provider backend ?
+
+For Ray worker node creation, each Ray worker node, we start a spark job with only one spark
+task to hold this Ray worker node. When we need to terminate this Ray worker node, we cancel
+the corresponding spark task so that the Ray worker node is terminated.
+
+The critical issue is, in the process that executing `NodeProvider` backends, how to get
+the spark session in parent process ? We can set up a spark job server that runs inside
```

Comment:
```suggestion
The critical issue is how to get the spark session in the NodeProvider process? We can set up a spark job server that runs inside
```

Reply:
I think we need to expand this to describe what the issue is related to spark session: like why we need to share the same spark session in different processes and what the solution is.

I saw you discussed why below. I think we can move to here.